TIM/TIM Core Code/man/man3/buffer_printf.3
.TH "buffer_printf" 3 "Tue Jan 26 2016" "Version 0.1" "P21451_TIM" \" -*- nroff -*-
.ad l
.nh
.SH NAME
buffer_printf \- printf Functionality for Buffer Module
.SS "Functions"

.in +1c
.ti -1c
.RI "void \fBPush_printf\fP (\fBbuffer_t\fP *buf, char *str,\&.\&.\&.)"
.br
.RI "\fIprintf implementation to char buffer \fP"
.ti -1c
.RI "void \fBPush_vprintf\fP (\fBbuffer_t\fP *buf, char *str, va_list vars)"
.br
.RI "\fIvprintf implementation to char buffer \fP"
.ti -1c
.RI "void \fBPushBinary16\fP (\fBbuffer_t\fP *buf, uint16_t x)"
.br
.RI "\fIPush 16 bit binary number to char buffer\&. \fP"
.ti -1c
.RI "void \fBPush_uint16\fP (\fBbuffer_t\fP *buf, uint16_t x)"
.br
.RI "\fIPush unsigned integer to char buffer\&. \fP"
.ti -1c
.RI "void \fBPush_int16\fP (\fBbuffer_t\fP *buf, int16_t x)"
.br
.RI "\fIPush integer to char buffer\&. \fP"
.ti -1c
.RI "void \fBPush_uint32\fP (\fBbuffer_t\fP *buf, uint32_t x)"
.br
.RI "\fIPush unsigned long integer to char buffer\&. \fP"
.ti -1c
.RI "void \fBPush_int32\fP (\fBbuffer_t\fP *buf, int32_t x)"
.br
.RI "\fIPush long to char buffer\&. \fP"
.ti -1c
.RI "void \fBPushStr\fP (\fBbuffer_t\fP *buf, char *str)"
.br
.RI "\fIPush char array (string) to char buffer\&. \fP"
.ti -1c
.RI "void \fBPushHex\fP (\fBbuffer_t\fP *buf, uint16_t x)"
.br
.RI "\fIPush 16 bit value to char buffer in hex format\&. \fP"
.ti -1c
.RI "void \fBPushFloat\fP (\fBbuffer_t\fP *buf, float x)"
.br
.RI "\fICheap implementation of float to char buffer\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
\fBbuffer_printf\&.h\fP
.PP
Created on: Mar 11, 2014 Author: Michael
.PP
\fBVersion:\fP
.RS 4
1\&.1 changed naming to match stdint and removed 2Buf from function names 
.PP
1\&.2 added l for int32 and b for 16-bit binary 
.RE
.PP

.SH "Function Documentation"
.PP 
.SS "void Push_int16 (\fBbuffer_t\fP * buf, int16_t x)"

.PP
Push integer to char buffer\&. Note this function is dependent on \fBPush_uint16()\fP
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print int to 
.br
\fIx\fP Integer to convert to text 
.RE
.PP

.PP
Definition at line 121 of file buffer_printf\&.c\&.
.SS "void Push_int32 (\fBbuffer_t\fP * buf, int32_t x)"

.PP
Push long to char buffer\&. Note this function is dependent on \fBPush_uint32()\fP
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print int to 
.br
\fIx\fP Long to convert to text 
.RE
.PP

.PP
Definition at line 129 of file buffer_printf\&.c\&.
.SS "void Push_printf (\fBbuffer_t\fP * buf, char * str,  \&.\&.\&.)"

.PP
printf implementation to char buffer Currently supports the following replacement flags:
.IP "\(bu" 2
\fCc\fP char, replaces flag with specified ASCII char
.IP "\(bu" 2
\fCb\fP binary, replaces flag with Binary representation See \fBPushBinary16()\fP
.IP "\(bu" 2
\fCl\fP long, replaces flag with Binary Representation See \fBPush_int32()\fP
.IP "\(bu" 2
\fCd\fP signed 16 bit integer, replaces flag with specified int\&. See \fBPush_int16()\fP
.IP "\(bu" 2
\fCe\fP \fCf\fP or \fCg\fP float, replaces flag with specified float See \fBPushFloat()\fP
.IP "\(bu" 2
\fCs\fP string, replaces flag with specified null terminated string See PushString()
.IP "\(bu" 2
\fCu\fP unsigned 16 bit integer, replaces flag with specified unsigned int See \fBPush_uint16()\fP
.IP "\(bu" 2
\fCx\fP 16 bit hex formated integer, replaces flag with 4 digit hex value See \fBPushHex()\fP
.PP
.PP
Example: 
.PP
.nf
1 buffer_t tx;
2 \&.\&.\&.
3 int16_t x = -1;
4 char name[] = "Muhlbaier";
5 Push_printf(&tx, "x = %d, hex - 0x%x, unsigned %u, name = %s);

.fi
.PP
 Would push to the buffer: 'x = -1, hex - 0xFFFF, unsigned 65535, name = Muhlbaier'
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print formatted string to 
.br
\fIstr\fP Pointer to null terminated string with replacement flags 
.br
\fI\&.\&.\&.\fP Variable argument list corresponding with replacement flags 
.RE
.PP

.PP
Definition at line 17 of file buffer_printf\&.c\&.
.SS "void Push_uint16 (\fBbuffer_t\fP * buf, uint16_t x)"

.PP
Push unsigned integer to char buffer\&. 
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print unsigned int to 
.br
\fIx\fP Unsigned integer to convert to text 
.RE
.PP

.PP
Definition at line 100 of file buffer_printf\&.c\&.
.SS "void Push_uint32 (\fBbuffer_t\fP * buf, uint32_t x)"

.PP
Push unsigned long integer to char buffer\&. 
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print unsigned long to 
.br
\fIx\fP Unsigned integer to convert to text 
.RE
.PP

.PP
Definition at line 137 of file buffer_printf\&.c\&.
.SS "void Push_vprintf (\fBbuffer_t\fP * buf, char * str, va_list vars)"

.PP
vprintf implementation to char buffer Same as \fBPush_printf()\fP except with a va_list pointer instead of an actual variable argument list\&. This allows other functions similar to Push_printf to be implemented\&.
.PP
For example: 
.PP
.nf
1 void LogStr(char * str, \&.\&.\&.) {
2    va_list vars;
3    va_start(vars, str);
4    // use Push_vprintf to log to SUBSYS_UART
5    Push_vprintf(SUBSYS_UART, str, vars);
6    va_end(vars);
7 }

.fi
.PP
.PP
See \fBPush_printf()\fP
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print formatted string to 
.br
\fIstr\fP Pointer to null terminated string with replacement flags 
.br
\fIvars\fP Variable argument list corresponding with replacement flags 
.RE
.PP

.PP
Definition at line 27 of file buffer_printf\&.c\&.
.SS "void PushBinary16 (\fBbuffer_t\fP * buf, uint16_t x)"

.PP
Push 16 bit binary number to char buffer\&. 
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print to 
.br
\fIx\fP Unsigned 16-bit integer to convert to text in binary format 
.RE
.PP

.PP
Definition at line 113 of file buffer_printf\&.c\&.
.SS "void PushFloat (\fBbuffer_t\fP * buf, float x)"

.PP
Cheap implementation of float to char buffer\&. Current implementation will format float as 0\&.000 by first printing out the integer portion of the float then multiplying the float by 1000 and subtracting the integer portion x 1000 and printing that after the \&.
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print float to 
.br
\fIx\fP Float value to convert to text 
.RE
.PP

.PP
Definition at line 171 of file buffer_printf\&.c\&.
.SS "void PushHex (\fBbuffer_t\fP * buf, uint16_t x)"

.PP
Push 16 bit value to char buffer in hex format\&. Will push four char's to the buffer, for example: A0F3
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print hex formatted int to 
.br
\fIx\fP Integer to convert to hex 
.RE
.PP

.PP
Definition at line 156 of file buffer_printf\&.c\&.
.SS "void PushStr (\fBbuffer_t\fP * buf, char * str)"

.PP
Push char array (string) to char buffer\&. 
.PP
\fBParameters:\fP
.RS 4
\fIbuf\fP Pointer to char buffer to print string to 
.br
\fIstr\fP Pointer to null terminated char array (e\&.g\&. string) 
.RE
.PP

.PP
Definition at line 150 of file buffer_printf\&.c\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for P21451_TIM from the source code\&.


TIM/TIM Core Code/buffer_printf.c
/*
 * buffer_printf.c
 *
 *  Created on: Mar 11, 2014
 *      Author: Michael
 */
#include "system.h"
#ifndef _BUFFER_PRINTF_H_
#error "buffer_printf.h must be included in system.h"
#endif
#ifndef _BUFFER_H_
#error "include buffer.h in system.h and buffer.c must be in your project"
#endif

#include <stdarg.h>

void Push_printf(buffer_t * buf, char * str, ...) {
    // variable argument list type
    va_list vars;
    // initialize the variable argument list pointer by specifying the
    // input argument immediately preceding the variable list
    va_start(vars, str);
    Push_vprintf(buf, str, vars);
    va_end(vars);
}

void Push_vprintf(buffer_t * buf, char * str, va_list vars) {
    char escape = 0;
    while (*str != 0) {
        if (*str == '%' && !escape) {
            switch (*(str + 1)) {
                case 0:
                    str++;
                    continue;
                case 'b':
                    // Function declared in Paterson_midterm.h/.c
                    PushBinary16(buf, va_arg(vars, unsigned int));
                    str += 2;
                    continue;
                case 'c': // char
                    // to be implemented by everyone else
#ifdef __XC32__
                    Push(buf, va_arg(vars, int));
#else
                    Push(buf, va_arg(vars, char));
#endif
                    str += 2;
                    continue;
                case 'd':
                    // Capo
                    // use va_arg to ge the next variable of length int
                    Push_int16(buf, va_arg(vars, int));
                    str += 2;
                    continue;
                case 'e':
                case 'f':
                case 'g': // scientific notation of float
                    // Jehandad
                    PushFloat(buf, va_arg(vars, double));
                    str += 2;
                    continue;
                case 'l':
                    Push_int32(buf, va_arg(vars, int32_t));
                    str += 2;
                    continue;
                case 's': // string
                    // Cecere / Paterson
                    PushStr(buf, va_arg(vars, char *));
                    str += 2;
                    continue;
                case 'u':
                    // use va_arg to ge the next variable of length int
                    Push_uint16(buf, va_arg(vars, unsigned int));
                    str += 2;
                    continue;
                case 'x': // hex
                    // Jake
                    PushHex(buf, va_arg(vars, int));
                    str += 2;
                    continue;
                default:
                    // if the character following the % was not "special"
                    // then we will print the % and whatever follows it
                    break;
            }
        }
        //escape = (*str == '\\') ? 1 : 0;
        //Push(buf,*str++);
        if (*str == '\\' && *(str + 1) == '%') {
            str++;
            escape = 1;
        } else {
            Push(buf, *str++);
            escape = 0;
        }
    }

}

void Push_uint16(buffer_t * buf, uint16_t x) {
    char num[6];
    unsigned int i = 0;
    do {
        num[i] = (x % 10) + '0';
        x /= 10;
        i++;
    } while (x);
    do {
        Push(buf, num[--i]);
    } while (i);
}

void PushBinary16(buffer_t * buf, uint16_t x) {
    uint16_t i;
    for (i = 0; i < 16; i++) {
        Push(buf, (x & 0x0001) ? '1' : '0');
        x >>= 1;
    }
}

void Push_int16(buffer_t * buf, int16_t x) {
    if (x < 0) {
        Push(buf, '-');
        x = -x;
    }
    Push_uint16(buf, x);
}

void Push_int32(buffer_t * buf, int32_t x) {
    if (x < 0) {
        Push(buf, '-');
        x = -x;
    }
    Push_uint32(buf, x);
}

void Push_uint32(buffer_t * buf, uint32_t x) {
    char num[10];
    unsigned int i = 0;
    do {
        num[i] = (x % 10) + '0';
        x /= 10;
        i++;
    } while (x);
    do {
        Push(buf, num[--i]);
    } while (i);
}

void PushStr(buffer_t * buf, char * str) {
    do {
        Push(buf, *str);
    } while (*str++ != 0);
}

void PushHex(buffer_t *buf, uint16_t x) {
    volatile unsigned int y;
    char num[4];
    unsigned int i = 0;
    for (i = 0; i < 4; i++) {
        y = x - ((x >> 4) << 4); // x % 16
        num[i] = (y >= 10) ? (y + 0x37) : (y + 0x30);
        x >>= 4; // x / 16
    } //while(x);

    do {
        Push(buf, num[--i]);
    } while (i);
}

void PushFloat(buffer_t * buf, float x) {
    int i;
    long l;
    // for now just print 0.000 format
    i = x;
    Push_int16(buf, i);
    l = (x * 1000);
    l = l - i * 1000;
    Push(buf, '.');
    if (l < 100) Push(buf, '0');
    if (l < 10) Push(buf, '0');
    if (l == 0) Push(buf, '0');
    else Push_uint16(buf, l & 0xFFFF);
}
